feat: parse inscription like witness data

bitSmiley opened the pull request:
# Description

<!--- Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

This is the start of a series of PRs that implements the solution mentioned in #2478. The implementation is broken down into several parts:
- [x] Extract memo from the witness script (This PR) 
- [ ] Detect targetted zeta btc address and extract corresponding witness script
- [ ] Javascript client that submits inscription transactions
- [ ] Integration testing in code base and in mainet

This PR adds a `DecodeScript` function that decodes memo wrapped in an inscription-like script.  The format of the script is following that of "inscription" defined in ordinal theory. However, to separate from inscription (as this use case is not an NFT), simplifications are made. The bitcoin envelope script is as follows:
```
OP_DATA_32 <32 byte of public key> OP_CHECKSIG
OP_FALSE
OP_IF
  OP_PUSH 0x...
  OP_PUSH 0x...
OP_ENDIF
```
The public key is the `internalKey` used in the taproot transaction. This follows `ord`.

The `DecodeScript` method take the script bytes and parse them according. Internally it is utilizing `ScriptTokenizer`. Ideally, it should directly use `txscript.ScriptTokenizer`, however, the current codebase does not have the versions for `ScriptTokenizer`. A simplified version is used. If `btcsuite` is upgraded, then `scriptTokenizer` can be removed and the code base will be simplified further.

// There are no content-type or any other attributes, it's just raw bytes.
# How Has This Been Tested?

<!--- Please describe the tests that you ran to verify your changes. Include instructions and any relevant details so others can reproduce. Link any optional github actions runs. -->

- [ ] Tested CCTX in localnet
- [ ] Tested in development environment
- [x] Go unit tests
- [ ] Go integration tests
- [ ] Tested via GitHub Actions


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **New Features**
	- Enhanced Bitcoin event processing with a new function to handle transactions containing witness data.
	- Added functionalities for decoding Bitcoin scripts, improving the robustness and validation of transaction scripts.

- **Tests**
	- Introduced a comprehensive test suite for the new script decoding functionality, validating various scenarios and error conditions.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

lumtis commented on zetaclient/chains/bitcoin/tx_script.go:
```diff
@@ -192,6 +193,36 @@ func DecodeOpReturnMemo(scriptHex string, txid string) ([]byte, bool, error) {
 	return nil, false, nil
 }
 
+// DecodeScript decodes memo wrapped in an inscription like script in witness
```
Not related to this PR directly but I would consider moving this entire file in an independent package: https://github.com/zeta-chain/node/issues/2525

But we can let as it in this PR to keep it minimal

swift1337 commented on zetaclient/chains/bitcoin/observer/inbound.go:
```diff
@@ -472,3 +472,39 @@ func GetBtcEvent(
+	)
+
+	if err != nil { // should never happen
+		return nil, err
```
```suggestion
		return nil, errors.Wrap(err, "unable to get btc event")
```

swift1337 commented on zetaclient/chains/bitcoin/observer/inbound.go:
```diff
@@ -472,3 +472,39 @@ func GetBtcEvent(
+// It will first prioritize OP_RETURN over tapscript.
+// The format of the tapscript is
+func GetBtcEventWithWitness(
+	rpcClient interfaces.BTCRPCClient,
```
```suggestion
	client interfaces.BTCRPCClient,
```

lumtis commented on zetaclient/chains/bitcoin/tx_script.go:
```diff
@@ -245,3 +276,211 @@
+//	 ...
+//		OP_PUSHDATA_N ...
+//
+// OP_ENDIF
+//
+// Note: the total data pushed will always be more than 80 bytes and within the btc transaction size limit.
+func decodeInscriptionPayload(t *scriptTokenizer) ([]byte, error) {
+	if !t.Next() || t.Opcode() != txscript.OP_FALSE {
+		return nil, fmt.Errorf("OP_FALSE not found")
+	}
+
+	if !t.Next() || t.Opcode() != txscript.OP_IF {
+		return nil, fmt.Errorf("OP_IF not found")
+	}
+
+	memo := make([]byte, 0)
+	var next byte
+	for {
+		if !t.Next() {
+			if t.Err() != nil {
+				return nil, t.Err()
+			}
+			return nil, fmt.Errorf("should contain more data, but script ended")
+		}
+
+		next = t.Opcode()
+
+		if next == txscript.OP_ENDIF {
+			break
+		}
+
+		if next < txscript.OP_DATA_1 || next > txscript.OP_PUSHDATA4 {
+			return nil, fmt.Errorf("expecting data push, found %d", next)
+		}
+
+		memo = append(memo, t.Data()...)
+	}
+
+	return memo, nil
+}
```
I think this suggestion makes sense

swift1337 commented on zetaclient/chains/bitcoin/observer/inbound.go:
```diff
@@ -472,3 +472,39 @@ func GetBtcEvent(
+// GetBtcEventWithWitness either returns a valid BTCInboundEvent or nil.
+// This method supports data with more than 80 bytes by scanning the witness for possible presence of a tapscript.
+// It will first prioritize OP_RETURN over tapscript.
+// The format of the tapscript is
```
Should this comment be extended?

bitSmiley replied:
oops, will update

lumtis commented on zetaclient/chains/bitcoin/tx_script.go:
```diff
@@ -245,3 +276,211 @@ func DecodeTSSVout(vout btcjson.Vout, receiverExpected string, chain chains.Chai
+// OP_PUSHBYTES_32 <32 bytes> OP_CHECKSIG <Content>
+func checkInscriptionEnvelope(t *scriptTokenizer) error {
+	if !t.Next() || t.Opcode() != txscript.OP_DATA_32 {
+		return fmt.Errorf("cannot obtain public key bytes")
+	}
+
+	if !t.Next() || t.Opcode() != txscript.OP_CHECKSIG {
+		return fmt.Errorf("cannot parse OP_CHECKSIG")
```
Maybe we can write the current opcode in the error?

lumtis commented on zetaclient/chains/bitcoin/tx_script.go:
```diff
@@ -245,3 +276,211 @@ func DecodeTSSVout(vout btcjson.Vout, receiverExpected string, chain chains.Chai
+	return nil
+}
+
+func makeScriptTokenizer(script []byte) scriptTokenizer {
```
```suggestion
func newScriptTokenizer(script []byte) scriptTokenizer {
```
I think new termninology is more adapted?

lumtis commented on zetaclient/chains/bitcoin/tx_script.go:
```diff
@@ -245,3 +276,211 @@ func DecodeTSSVout(vout btcjson.Vout, receiverExpected string, chain chains.Chai
+
+		return true
+	case op > txscript.OP_PUSHDATA4:
+		t.err = fmt.Errorf("unexpected op code")
```
Same, could we add the value of the opcode here?

swift1337 commented on zetaclient/chains/bitcoin/tx_script.go:
```diff
@@ -192,6 +193,36 @@ func DecodeOpReturnMemo(scriptHex string, txid string) ([]byte, bool, error) {
+	t := makeScriptTokenizer(script)
+
+	if err := checkInscriptionEnvelope(&t); err != nil {
+		return nil, false, err
```
```suggestion
		return nil, false, errors.Wrap(err, "unable to check the envelope")
```

lumtis reviewed (commented):
>This is the start of a series of PRs that implements the solution mentioned in https://github.com/zeta-chain/node/issues/2478. The implementation is broken down into several parts:

Sounds good. For task that can be broken down in subset, we usually create issue tagged as `epic`

Example https://github.com/zeta-chain/node/issues/2506

swift1337 commented on zetaclient/chains/bitcoin/tx_script.go:
```diff
@@ -245,3 +276,211 @@ func DecodeTSSVout(vout btcjson.Vout, receiverExpected string, chain chains.Chai
+	return nil
+}
+
+func makeScriptTokenizer(script []byte) scriptTokenizer {
```
can we move this to `zetaclient/chains/bitcoin/tokenizer`?

`t := tokenizer.New(script) *Tokenizer`

bitSmiley replied:
Actually it would be great to update the btcsuite, then this code can be deleted, haha

swift1337 commented on zetaclient/chains/bitcoin/tx_script_test.go:
```diff
@@ -491,3 +491,66 @@ func TestDecodeTSSVoutErrors(t *testing.T) {
+func TestDecodeScript(t *testing.T) {
+	t.Run("should decode longer data ok", func(t *testing.T) {
+		// 600 bytes of random data generated offline
+		data := "2001a7bae79bd61c2368fe41a565061d6cf22b4f509fbc1652caea06d98b8fd0c7ac00634d0802c7faa771dd05f27993d22c42988758882d20080241074462884c8774e1cdf4b04e5b3b74b6568bd1769722708306c66270b6b2a7f68baced83627eeeb2d494e8a1749277b92a4c5a90b1b4f6038e5f704405515109d4d0021612ad298b8dad6e12245f8f0020e11a7a319652ba6abe261958201ce5e83131cd81302c0ecec60d4afa9f72540fc84b6b9c1f3d903ab25686df263b192a403a4aa22b ...
```
can we generate this data on the fly? Also would be cool to have some real-world use-case of having the following logic:
```
Given an ABI
Encode some call data using inscriptions 
Assert memo 
Make sure that passing memo further allows proper ABI decoding and calling

bitSmiley replied:
In the following PR will contain parsing the txn data to extract the witness. Then a real mainet txn will be tested. There will be a nodejs repo for the use case you mentioned. Regarding the generate this data on the fly, might need to look into the existing library script building capability, currently it's generated using nodejs.

swift1337 reviewed (commented):
Thank you for contributing! Looks awesome, left some comments.

bitSmiley commented:
@lumtis @swift1337 I have addressed the review feedbacks. I'm not familiar with the "resolve" convention you have, so I just leave them open first. Please take another look and let me know.

lumtis commented on zetaclient/chains/bitcoin/observer/inbound.go:
```diff
@@ -472,3 +472,39 @@ func GetBtcEvent(
+		return event, nil
+	}
+
+	// TODO: integrate parsing script
```
@coderabbitai yes, can you create an issue for this and suggest a comment line here to link the issue?

lumtis reviewed (commented):
Looks like we need to run
```
make generate
```
in the repo and commit the changes

bitSmiley commented:
@lumtis Ran `make generate` and pushed the results

fbac commented on zetaclient/chains/bitcoin/observer/inbound.go:
```diff
@@ -477,3 +477,38 @@ func GetBtcEvent(
+		netParams,
+		depositorFee,
+	)
+
+	if err != nil { // should never happen
+		return nil, errors.Wrap(err, "unable to get btc event")
+	}
+	if event != nil {
+		return event, nil
+	}
```
nit: just identation changes. The comment does not add a lot of value, so we can skip it.
```suggestion

	if err != nil {
		return nil, errors.Wrap(err, "unable to get btc event")
	}

	if event != nil {
		return event, nil
	}
```

fbac commented on zetaclient/chains/bitcoin/tokenizer.go:
```diff
@@ -0,0 +1,156 @@
+	"github.com/btcsuite/btcd/txscript"
+)
+
+func newScriptTokenizer(script []byte) scriptTokenizer {
```
Does `script` have a min or max bytes size that we should be checking here?

bitSmiley replied:
It's coming from Bitcoin, it should be contained. I think we can just take whatever comes from BTC.

fbac replied:
Makes sense, I'm not familiar with this scripts so that's why I was wondering if there's a possibility of script being 0 bytes, on which case we can simply discard it and save resources.

fbac commented on zetaclient/chains/bitcoin/tokenizer.go:
```diff
@@ -0,0 +1,156 @@
+		script := t.script[t.offset:]
+
+		// add 2 instead of 1 because script includes the opcode as well
+		length := int(op) - txscript.OP_DATA_1 + 2
```
question: to understand this line, why OP_DATA_1 has to be subtracted?

It seems like OP_DATA_1 has value 1, so we're doing `X - 1 + 2` instead of simply `X + 1`? That would give us always the script length plus the opcode?

Not sure if I'm missing something.

bitSmiley replied:
Yeah, that makes sense. When this was written, I was thinking it might be easier to interpret. I will update accordingly and put in comments.

fbac commented on zetaclient/chains/bitcoin/tokenizer.go:
```diff
@@ -0,0 +1,156 @@
+		// Move the offset forward and set the opcode and data accordingly.
+		t.offset += length
+		t.op = op
+		t.data = script[1:length]
+
+		return true
+	case op > txscript.OP_PUSHDATA4:
+		t.err = fmt.Errorf("unexpected op code %d", op)
+		return false
+	// Data pushes with parsed lengths -- OP_PUSHDATA{1,2,4}.
+	default:
```
Just some linting.
```suggestion
		t.data = script[1:length]
		return true

	case op > txscript.OP_PUSHDATA4:
		t.err = fmt.Errorf("unexpected op code %d", op)
		return false

	// Data pushes with parsed lengths -- OP_PUSHDATA{1,2,4}.
	default:
```

fbac commented on zetaclient/chains/bitcoin/tokenizer.go:
```diff
@@ -0,0 +1,156 @@
+		// add 2 instead of 1 because script includes the opcode as well
+		length := int(op) - txscript.OP_DATA_1 + 2
+		if len(script) < length {
+			t.err = fmt.Errorf("opcode %d requires %d bytes, but script only "+
+				"has %d remaining", op, length, len(script))
```
nit: 
Consider shortening and making the error more specific, as it's repeated in the code. Concatenating strings is not optimal.
```suggestion
			t.err = fmt.Errorf("opcode %d detected, but script only %d bytes remaining", op, len(script))
```

fbac commented on zetaclient/chains/bitcoin/tokenizer.go:
```diff
@@ -0,0 +1,156 @@
+
+		script := t.script[t.offset+1:]
+		if len(script) < length {
+			t.err = fmt.Errorf("opcode %d requires %d bytes, but script only "+
```
Same here, avoid concatenating.

fbac commented on zetaclient/chains/bitcoin/tokenizer.go:
```diff
@@ -0,0 +1,156 @@
+	// Data pushes with parsed lengths -- OP_PUSHDATA{1,2,4}.
+	default:
+		var length int
+		switch op {
+		case txscript.OP_PUSHDATA1:
+			length = 1
+		case txscript.OP_PUSHDATA2:
+			length = 2
+		default:
+			length = 4
+		}
```
Default to a defensive approach:

```suggestion
		switch op {
		case txscript.OP_PUSHDATA1:
			length = 1
		case txscript.OP_PUSHDATA2:
			length = 2
		case txscript.OP_PUSHDATA4:
		        length = 4
		default:
			return fmt.Errorf("unexpected op code %d", op)
		}
```

fbac commented on zetaclient/chains/bitcoin/tokenizer.go:
```diff
@@ -0,0 +1,156 @@
+		switch length {
+		case 1:
+			dataLen = int(script[0])
+		case 2:
+			dataLen = int(binary.LittleEndian.Uint16(script[:2]))
+		case 4:
+			dataLen = int(binary.LittleEndian.Uint32(script[:4]))
```
Avoiding magic numbers when possible.

```suggestion
		case 2:
			dataLen = int(binary.LittleEndian.Uint16(script[:length]))
		case 4:
			dataLen = int(binary.LittleEndian.Uint32(script[:length]))
```

fbac commented on zetaclient/chains/bitcoin/tokenizer.go:
```diff
@@ -0,0 +1,156 @@
+
+		// Disallow entries that do not fit script or were sign extended.
+		if dataLen > len(script) || dataLen < 0 {
+			t.err = fmt.Errorf("opcode %d pushes %d bytes, but script only "+
```
Same consideration on the error as before.

fbac commented on zetaclient/chains/bitcoin/tokenizer.go:
```diff
@@ -0,0 +1,156 @@
+		}
+
+		// Move the offset forward and set the opcode and data accordingly.
+		t.offset += 1 + length + dataLen
```
question: is `1` the opcode byte? should we `t.offset += int(op) + length + dataLength`?

bitSmiley replied:
1 is the opcode size, which is just 1 byte. `int(op)` is the opcode value, it should not be mixed with the size.

fbac replied:
Oh yes, thank you I confused it here, definitely not using `int(op)` here. In that case could you extend the comment and mention 1 refers to the 1 byte used by the opcode?

fbac commented on zetaclient/chains/bitcoin/tokenizer.go:
```diff
@@ -0,0 +1,156 @@
+
+// Done returns true when either all opcodes have been exhausted or a parse
+// failure was encountered and therefore the state has an associated error.
+func (t *scriptTokenizer) Done() bool {
```
Consideration for all the scriptTokenizer methods: are we going to use them outside the `bitcoin` package? If that's not the case we can simply make them private.

bitSmiley replied:
I would say keep it as it is for now. Ideally we should be using the bitcoin script functionalities from btcsuite. The current version in the code base does not have these functionalities. But higher versions do. In this case, we should keep the method name as of now and once versions are upgraded, just swap to the library implementation.

fbac replied:
Makes sense, thanks!

fbac commented on zetaclient/chains/bitcoin/tx_script.go:
```diff
@@ -245,3 +275,44 @@ func DecodeTSSVout(vout btcjson.Vout, receiverExpected string, chain chains.Chai
+		return nil, fmt.Errorf("OP_IF not found")
+	}
+
+	memo := make([]byte, 0)
```
What is the maximum byte size a bitcoin message can carry?

bitSmiley replied:
From what I know, taproot script txn size is lifted, so to play safe we assume the max block size, 4MB. But usually a few hundred KB is already very expensive.

bitSmiley commented:
@fbac I have updated based on your feedbacks. Please help take a look.

fbac commented:
> @fbac I have updated based on your feedbacks. Please help take a look.

Thank you for the changes, it looks good to me now. Could you fix the two errors concatenating strings and expand the comment where we add 1? After that this can be merged!

bitSmiley commented:
@fbac Updated so. Funny thing, seems like my IDE added that concatenation, fix that already.

fbac reviewed (approved):
LGTM, thank you!

fbac commented:
Rebasing the branch with main, if we pass all the checks this will be merged.

Also @bitSmiley could you update the changelog introducing this change?
